### DISCUSSION TITLE
Integrate with Kubernetes

### DISCUSSION THREAD

**Discussion by mraerino** [testcontainers/testcontainers-go#510] [Ideas]
Someone at a meetup lately suggested it would be cool to be able to start containers on Kubernetes so blackbox tests of containers can be done using standard test libraries.

Since this library is in an early state, it would be a great opportunity to abstract the container layer in such a way that it does not matter whether the Docker or Kubernetes APIs are used.

**Comment by gianarb** [comment]
Yep, this is a good idea, I think we should create a Container interface with the current functions we implemented in the container like:

```go
type Container interface{
     Terminate(ctx context.Context, t *testing.T) error
     LivenessCheckPorts(ctx context.Context) (nat.PortSet, error) 
     GetIPAddress(ctx context.Context) (string, error)
     GetHostEndpoint(ctx context.Context, port string) (string, string, error)
}
```
And we can have a new file `kubernetes.go` that implements that interface with a function `RunPod` (it should return a container just like [`RunContainer`](https://github.com/testcontainers/testcontainer-go/blob/master/docker.go#L109) but via kubernetes (a pod).

Nice idea! I don't think this is a top priority btw. For example, I am using this library in a project and we use Jenkins to run tests against kubernetes but we use Docker in Docker. I am not sure about the benefits of having the ability to run pods directly. What do you think?

**Comment by mraerino** [comment]
That sounds great!

**Comment by mraerino** [comment]
The foundation for this has been laid in #36

**Comment by rnorth** [comment]
This is probably something we should do with the Java implementation and other forks, at the right time. Please could we discuss the general approach to tackling this?

I can't think of anything specific, but it would be good for all the implementations to be gradually moving towards convergence on feature sets rather than divergence 😄 .

**Comment by gianarb** [comment]
We have a lot to fix before this. So yes! Let's work on this together when
everyone will be ready.

**Comment by mdelapenya** [comment]
I think it makes sense to move this thread to Discussions

**Comment by prskr** [comment]
Hey, not sure if this still 'a thing' but could you probably elaborate how you imagine the workflow or what could be the benefit?

- being able to run tests within a Kubernetes cluster by deploying test services as Deployments+Service vs. local containers? (I'm assuming here testcontainers should probably provide both the Deployment/StatefulSet __and__ the Service)
- Using 'testcontainers' to run integration tests of your packages e.g. from a CI pipeline probably running outside of the cluster by using the default deployment mechanism (e.g. Helm, Kustomize, Tanka, ... )?

I'm just wondering because the former approach could be done rather easy and could be really helpful because Docker-in-Docker is no fun in Kubernetes :smile: but the latter approach could be rather hard because there are so many different deployment mechanisms and what about CRDs and stuff...?

**Comment by kiview** [comment]
A lot of the reasons why we think that k8s integration as envisioned in most cases does not really make sense for the concept of Testcontainers have been outlined in this issue in the `tc-java` repository:
https://github.com/testcontainers/testcontainers-java/issues/1135#issuecomment-453757407
